### ISSUE TITLE
[v0.2.16]MoE

### ISSUE THREAD

**Issue by m96-chan** [m96-chan/PyGPUkit#110]
![Image](https://github.com/user-attachments/assets/e7d4c332-c4ae-41f2-b684-1fd70544a634)

## Summary

Add Mixture of Experts (MoE) support for models like Mixtral, Qwen-MoE, DeepSeek-V2/V3, DBRX.

## Target Models

| Model | Experts | Top-K | Architecture |
|-------|---------|-------|--------------|
| Mixtral 8x7B | 8 | 2 | Standard MoE |
| Qwen-MoE | 60 | 4 | Fine-grained MoE |
| DeepSeek-V2 | 160 | 6 | Fine-grained + shared expert |
| DeepSeek-V3 | 256 | 8 | Multi-head Latent Attention + MoE |
| DBRX | 16 | 4 | Fine-grained MoE |

### 1. Router/Gating Network
- [ ] `moe_gate` - Linear projection: `[batch, seq, hidden] -> [batch, seq, num_experts]`
- [ ] Softmax over experts dimension

### 2. Top-K Expert Selection
- [ ] `topk_with_indices` - Returns `(values, indices)` for top-k experts per token
  - Input: `[batch, seq, num_experts]`
  - Output: `([batch, seq, k], [batch, seq, k])`
- [ ] Efficient GPU implementation (radix select or partial sort)

### 3. Token Dispatch (Scatter)
- [ ] `moe_permute` / `moe_dispatch` - Route tokens to selected experts
  - Reorder tokens by expert assignment
  - Build expert offset table for variable batch sizes
- [ ] Support for capacity factor (drop tokens if expert overloaded)

### 4. Expert FFN Computation
- [ ] `grouped_gemm` - Batch GEMM with variable group sizes
  - Each expert processes different number of tokens
  - CUTLASS GroupedGEMM or custom kernel
- [ ] Shared expert support (DeepSeek-V2/V3 style)

### 5. Token Combine (Gather + Weighted Sum)
- [ ] `moe_unpermute` / `moe_combine` - Combine expert outputs
  - Weighted sum by router probabilities
  - Restore original token order

### 6. Auxiliary Components
- [ ] Load balancing loss (for training/fine-tuning)
- [ ] Expert capacity management
- [ ] Token dropping strategy

### Phase 1: Basic MoE (Mixtral compatible)
```
tokens -> gate -> topk -> permute -> experts(grouped_gemm) -> unpermute -> output
```

### Phase 2: Fine-grained MoE (Qwen/DeepSeek)
- Larger number of smaller experts
- Shared expert pathway
- More sophisticated load balancing

### Phase 3: Performance Optimization
- Fused kernels (gate + topk + permute)
- Expert parallelism across GPUs
- Paged expert cache

## Kernel Priorities

| Priority | Kernel | Difficulty | Notes |
|----------|--------|------------|-------|
| P0 | `topk_with_indices` | Medium | Core routing operation |
| P0 | `moe_permute` | Medium | Token reordering |
| P0 | `grouped_gemm` | High | Variable batch GEMM |
| P0 | `moe_unpermute` | Medium | Output combination |
| P1 | Fused gate+topk | Medium | Reduce memory traffic |
| P1 | Expert parallelism | High | Multi-GPU support |

## Reference Implementations

- [Megablocks](https://github.com/databricks/megablocks) - Efficient MoE training
- [vLLM MoE](https://github.com/vllm-project/vllm/tree/main/csrc/moe) - Inference-focused
- [CUTLASS GroupedGEMM](https://github.com/NVIDIA/cutlass/tree/main/examples/24_gemm_grouped)
- [SGLang MoE](https://github.com/sgl-project/sglang) - FP8 MoE support

## File Structure

```
native/ops/moe/
  moe.cu              # Dispatch and kernel launch
  moe_kernels.cuh     # Core MoE kernels
  topk_kernels.cuh    # Top-K selection
  permute_kernels.cuh # Token routing
  grouped_gemm.cuh    # Variable batch GEMM

src/pygpukit/ops/
  moe.py              # Python API

src/pygpukit/llm/layers/
  moe_layer.py        # MoE layer implementation
```

## Success Criteria

- [ ] Mixtral 8x7B inference working
- [ ] Performance within 20% of vLLM MoE
- [ ] Support for FP16/BF16 (FP8 future)
- [ ] Memory efficient (no full expert activation)
